data/transactions/logic/TEAL_opcodes.md

```diff
@@ -769,6 +769,7 @@ Decodes A using the base64 encoding E. Specify the encoding with an immediate ar
 - Opcode: 0x5d {string return type}
 - Stack: ..., A: []byte, B: []byte &rarr; ..., any
 - return key B's value from a [valid](jsonspec.md) utf-8 encoded json object A
+- **Cost**: 25 + 2 per 7 bytes
```

Comment:
2 per 7 bytes is a weird combination. The actual linear slope for each byte is 0.271 so the contenders were 1/3, 2/7, 3/11, etc. Ideally we want a simple enough fraction which is as close to the actual cost per byte as possible since even an 0.05 difference in a 1000 character long JSON bytestring ends up being a cost difference of about 50. This immediately excluded 1/3 and so 2/7 was the next best option. However, 3 per 11 bytes could make more sense for this.

Reply:
I like the idea but I'm not sure it would work in practice because the edge cases are so much larger than what we're imagining the average JSON string length to be.

I'm imagining a 2000-char-long JSON string edge case and a ~100-long expected case. Let's assume for now that the 2000-char-long string is the largest possible string. The current model would cost the 100-long input 53 and the 2000-long input 596.

If we were to scale the linear equation by shifting the y-intercept, then the new y-intercept would need to be 96. This would mean that the average 100-long input would now cost 121 – more than double what it would cost previously.

Reply:
I believe this is outdated. Resolving.

Reply:
so if we start with y-intercept of 50 then for `y = 50 + x / 4` we get:
* 100-long: 75
* 2000-long: 550
That seems not too far off from the model. Op code costs have a lot of variability in practice so it really depends what the unit of comparison is. We want to discourage DOS'ing so it's important to look at the max length (4096)

Reply:
Given what I said, here are the calcs for 4096:
* current model: 1196 (assuming round up)
* 50 + x/4 model: 1029

Reply:
That's a fair point that op codes don't have to be entirely exact. We could probably get away with using 50 and a more sensible fraction of 1/4. Thank you for your explanations.

One potential issue I worry about would be the case in which you are trying to extract a nested key. Suppose there is a JSON object of length 100 with an inner object of length 50 and an inner object inside that one of length 10. Accessing a value in that inner inner object would require 3 json_ref calls, so the increased base cost would have a larger impact on the overall cost of accessing nested keys.

Using the values from above, the 25 + 2/7x model would cost 120 and the 50 + 1/4x one would cost 190. That being said, the differences would become negligible as the lengths became larger; it would more negatively affect smaller strings.

The answer could just be that we accept this added base cost. On the other hand, it does raise the question of how important it is for the cost numbers to be "nice".

Reply:
I don't know what the right answer is. If we had usage patterns we could get better estimates. Philosophically, I'd rather be conservative starting out the gate as if we're too conservative, we can loosen the requirements without much negative impact. But if we end up charging too little, then it's hard to tighten things up going forward. Of course, the cost shouldn't be so overly conservative so as to completely shut off op usage.